### PR TITLE
Implemented product-list page project

### PR DESCRIPTION
**Pull request by volodymyr35**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
   },
   "dependencies": {
     "react": "^18.2.0",
-    "react-dom": "^18.2.0"
+    "react-dom": "^18.2.0",
+    "react-router-dom": "^6.22.2"
```
Comment (Hollister009) [inline package.json]: Why do we need `react-router-dom` here?

File: src/App.jsx [hunk 2]
```diff
@@ -1,35 +1,85 @@
...
+import { useState } from "react";
+import Footer from "./components/Footer/Footer";
+import Header from "./components/Header/Header";
+import ItemList from "./components/ItemList/ItemList";
+import data from "./data/pdp.json";
+import Cart from "./components/Cart/Cart";
 
 function App() {
-  const [count, setCount] = useState(0)
+  const [cartState, setCartState] = useState([]);
```
Comment (Hollister009) [inline src/App.jsx]: I think we can move state into it's separate hook/context

File: src/App.jsx [hunk 3]
```diff
@@ -1,35 +1,85 @@
...
+  const [wishlist, setWishlist] = useState([]); // Array of IDs (Array<string>)
+  const [isCartOpen, setCartOpen] = useState(false);
+
+  const updateItemQuantityById = (id, quantity) => {
+    const currentItemIndex = cartState.findIndex((item) => item.id === id);
```
Comment (Hollister009) [inline src/App.jsx]: Here, we can utilize the logic from `lodash/xorBy`

File: src/components/Cart/Cart.css [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+.cart__item-list {
+  display: flex;
+  flex-direction: column;
+  justify-content: space-between;
+  height: auto;
+  overflow-y: auto;
+  width: 80%;
+}
+
+/* .cart__item {
```
Comment (Hollister009) [inline src/components/Cart/Cart.css]: No need to have the commented out code

File: src/components/Cart/Cart.jsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+}) {
+  const purchaseHandler = () => {
+    clearCart();
+
+    setTimeout(() => {
+      alert("Thank you for your purchase!");
+    });
+  };
+
+  const cartTotalPrice = state.reduce(
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx]: This needs to be memoized

File: src/components/Cart/Cart.jsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    0
+  );
+
+  if (!isOpen) return null;
+
+  if (state.length === 0) {
+    return (
+      <div className="cart__container">
+        <div className="cart__buttons">
+          <button className="purchase__button" disabled>
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx]: Why do we need separate classes for purchase and close buttons?

File: src/components/Cart/Cart.jsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
+/* eslint-disable react/no-unknown-property */
+/* eslint-disable react/prop-types */
+import cart from "../../images/cart.svg";
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx]:
```suggestion
import { ReactComponent as CartIcon } from "../../images/cart.svg";
```

File: src/components/Cart/Cart.jsx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      <div className="cart__container">
+        <div className="cart__buttons">
+          <button className="purchase__button" disabled>
+            Purchase
+          </button>
+          <button className="close__button" onClick={handleClose}>
+            X
+          </button>
+        </div>
+        <img src={cart} alt="cart" width="40" />
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx]: BTW we can use svg as React components:
https://create-react-app.dev/docs/adding-images-fonts-and-files/#adding-svgs

File: src/components/Cart/Cart.jsx [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+          <CartItem
+            key={item.id}
+            {...item}
+            updateItem={updateItem}
+            removeItem={removeItem}
+          />
+        ))}
+      </ul>
+      <span className="total">
+        Total price: {formatCurrency.format(cartTotalPrice)}
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx:77]: The total-price value must be memoized

File: src/components/Cart/Cart.jsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      <div className="cart__buttons">
+        <button className="purchase__button" onClick={purchaseHandler}>
+          Purchase
+        </button>
+        <button className="close__button" onClick={handleClose}>
+          X
+        </button>
+      </div>
+      <ul className="cart__item-list">
+        {state.map((item) => (
```
Comment (Hollister009) [inline src/components/Cart/Cart.jsx]: Can be extracted into its own function

File: src/components/Cart/CartItem.jsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        <span>
+          Quantity:
+          <input
+            id="item__quantity"
+            className="count__quantity"
+            type="number"
+            value={quantity}
+            min={MIN_QUANTITY}
+            max={MAX_QUANTITY}
+            onChange={(e) => {
```
Comment (Hollister009) [inline src/components/Cart/CartItem.jsx]: This should be extracted into it's own callback

File: src/components/Counter/Counter.jsx [hunk 12]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable react/prop-types */
+import './Counter.css';
+
+function Counter({count = 0}) {
+  if(count === 0) return null;
+
+  return (
+    <div className="counter">
+    <span className="counter__count">{count}</span>
+  </div>
```
Comment (Hollister009) [inline src/components/Counter/Counter.jsx]: This file needs to be formatted

File: src/components/Footer/Footer.jsx [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+function Footer() {
+  return (
+    <footer className='footer'>
+      <p className="footer_copyrights">
+       <span> &copy; 2024 </span>
+      </p>
+    </footer>
+  )
+
```
Comment (Hollister009) [inline src/components/Footer/Footer.jsx]: Remove this line

File: src/components/Header/Header.jsx [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import Counter from "../Counter/Counter";
+
+import "./Header.css";
+
+function Header({ wishlist = [], cartState = [], handleCartOpen }) {
+
+  const cartItemCount = cartState.reduce((acc, {quantity}) => acc + quantity, 0);
+  
+  return (
+    <div className="header">
```
Comment (Hollister009) [inline src/components/Header/Header.jsx:18]: Please fix formatting

File: src/components/Header/Header.jsx [hunk 15]
```diff
@@ -0,0 +1,34 @@
+/* eslint-disable react/prop-types */
+import cart from "../../images/cart.svg";
+import heart from "../../images/heart.svg";
+import Counter from "../Counter/Counter";
+
+import "./Header.css";
+
+function Header({ wishlist = [], cartState = [], handleCartOpen }) {
+
+  const cartItemCount = cartState.reduce((acc, {quantity}) => acc + quantity, 0);
```
Comment (Hollister009) [inline src/components/Header/Header.jsx]: useMemo here

File: src/components/Item/Item.jsx [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import cart from "../../images/cart.svg";
+import heart from "../../images/heart.svg";
+import heartFilled from "../../images/heart-black.svg";
+
+import "./Item.css";
+
+export const MAX_QUANTITY = 31;
+export const MIN_QUANTITY = 1;
+
+const formatCurrency = new Intl.NumberFormat(undefined, {
```
Comment (Hollister009) [inline src/components/Item/Item.jsx:9]: Needs to be extracted into constant

File: src/context/appContext.jsx [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+    setCart(newCart);
+  };
+
+  const clearCart = () => {
+    setCart([]);
+  }
+
+  return (
+    <AppContext.Provider value={{ cartData, addToCart, removeFromCart, clearCart }}>
```
Comment (Hollister009) [inline src/context/appContext.jsx]: Extract value:
`const values = useMemo(() => ({ cartData, addToCart, removeFromCart, clearCart }), [cartData, addToCart, removeFromCart, clearCart])`

File: package.json (hunk 1 above)
Comment (volodymyr35) [reply package.json]: Done!

File: src/components/Cart/Cart.jsx (hunk 6 above)
Comment (volodymyr35) [reply src/components/Cart/Cart.jsx]: Done!

File: src/components/Cart/CartItem.jsx (hunk 11 above)
Comment (volodymyr35) [reply src/components/Cart/CartItem.jsx]: Done!

File: src/components/Footer/Footer.jsx (hunk 13 above)
Comment (volodymyr35) [reply src/components/Footer/Footer.jsx]: Done!

**Review by Hollister009 — APPROVED** [review]
LGTM!

**Pull request merged**